File: cpp/006_ZigZagConversion.cpp
class Solution {
public:
    string convert(string s, int nRows) {
        if(nRows <= 1) return s;
        if(s.length() <= 2) return s;
        string tmp[nRows];
        int i = 0;
        int len = s.length();
        int zag = nRows - 2;
        while( i < len ){
            for( int j = 0; i < len && j < nRows; ++j ){
                tmp[j] += s[i];
                ++i;
            }
            for( int k = zag; i < len && k > 0; --k){
                tmp[k] += s[i];
                ++i;
            }
        }
        string res;
        for( int i = 0; i < nRows; ++i){
            res += tmp[i];
        }
        return res;
    }
};

File: js/006_ZigZagConversion.js
/**
 * @param {string} s
 * @param {number} numRows
 * @return {string}
 */
var convert = function(s, numRows) {
    if( numRows <= 1 ) return s;
    if( s.length <= 2 ) return s;
    var tmp = new Array(numRows);
    var i = 0;
    var len = s.length;
    var zag = numRows - 2;
    for( var n = 0; n < len; ++n){
        tmp[n] = "";
    }
    while( i < len ){
        for( var j = 0; i < len && j < numRows; ++j ){
            tmp[j] += s[i];
            ++i;
        }
        for( var k = zag; i < len && k > 0; --k ){
            tmp[k] += s[i];
            ++i;
        }
    }
    var res = "";
    for( var c in tmp){
        res += tmp[c];
    }
    return res;
};